rfcs/20200106-prelu/README.md

```diff
@@ -0,0 +1,137 @@
+``` cpp
+#define DNNL_ARG_WEIGHTS number
+#define DNNL_ARG_DIFF_WEIGHTS number
+```
+Examples:
+n = 256, h = 7, w = 7 c=128
+m_data = {{256, 7, 7, 128}, f32, nhwc};
+
+Algorithm:
+- whole-tensor (tensorflow wariant) - weights_data = {{256, 7, 7, 128}, f32, nhwc};
+
+### Broadcast semantic
+
+Specifying format tag "any" for the weights tensor will enable broadcast feature.
+It's purpose is to provide support for different algorithms: channel_wise,
+channel_shared, shared_axes. Weights will be broadcasted to src tensor if they
+will be broadcastable (see definition below). Format tag of weights will be
+deduced as src data format. Data type of weights will be preserved.
+
+Definition of broadcastable weights (please consider cpp example as pseudocode):
+``` cpp
+bool are_weights_brodcastable(const dims_t &src, const dims_t &weights) {
+
+    if (src.size() != weights.size())
+        return false;
+
+    for (std::size_t i = 0; i < src.size(), i++)
+        if (weights[i] !=1 && weights[i] != src[i])
+            return false;
+    return true;
+}
+```
+Examples:
+n = 256, h = 7, w = 7 c=128
+m_data = {{256, 7, 7, 128}, f32, nhwc};
+- weights_data = {{1, 1, 1, 128}, f32, any} // is_broadcastable = true,
+channel_wise;
+- weights_data = {{1, 1, 1, 1}, f32, any} // is_broadcastable = true;
+channel_shared;
+- weights_data = {{256, 1, 1, 128}, f32, any} // is_broadcastable = true;
+shared_axes (tf/keras)
+- weights_data = {{16, 3, 7, 128}, f32, any} // is_broadcastable = false;
+- weights_data = {{1}, f32, any} // is_broadcastable = false;
+
+2. Option B - extend exisiting eltwise implementation
+
+I would recommend option number A. In case of calculation gradients for alpha we
+ must pass memory descriptors for alpha and alpha_grad. Implementation
+ in existing eltwise kernel would require breaking the ABI. What is more now in
+ eltwise we pass float as alpha param. If we added DNNL_ARG_ALPHA required for
+ PReLU and alpha required for LeakyReLU remained it would be ambiguous for the
+ user.
+
+## Open questions
+1. Tensorflow gives user greater flexibility than channel wise/shared,
+by shared_axes parameter. Should we support that as well ?
+
+API wise yes, implementation of full functionality can be spread over time
+according to the internal priorities.
+
+2. Should other types of broadcast be supported?
+
+Example:
+
+Changed definition of broadcastable weights:
+``` cpp
+bool are_weights_brodcastable(const dims_t &src, const dims_t &weights) {
+
+    if (src.size() != weights.size())
+        return false;
+
+    for (std::size_t i = 0; i < src.size(), i++)
+        if (src[i] % weights[i] != 0)
+            return false;
+    return true;
+}
+```
+After changed definition following case will be allowed:
+n = 256, h = 7, w = 7 c=128
+src_data = {{256, 7, 7, 128}, f32, nhwc};
+- weights_data = {{128, 1, 1, 128}, f32, any}
+
+Currently I don't see framework that support that, but technically it is doable.
```

Comment:
No. I don't think even NumPy supports such semantics. Given that there are other primitives that support broadcasting (matmul, binary), this option if considering, is a completely separate RFC. So I suggest drop this altogether from here.

Reply:
Ok.